Don't treat no changes as an error

iconara opened the pull request:
When the change set contains no changes CFN returns an error. However, this shouldn't be an error state in Cuffsert.

When scripting with Cuffsert it complicates things a lot that no changes results in the tool exiting with a non-zero value. Either you have to ignore errors completely, or capture the output of the command and look for the "didn't contain changes" string – which is no fun at all in Bash.

I haven't been able to think of a situation where I want Cuffsert to exit with a non-zero value when there are no changes. If someone can come up with one it would probably be possible to add a flag that enabled a mode that treated no changes as an error state – but I would argue that the change I'm proposing here should be the default and that flag should be added when needed.

iconara committed:
Don't treat no changes as an error

When the change set contains no changes CFN returns an error. However, this is not an error state. When scripting with Cuffsert it complicates things a lot that no changes results in the tool exiting with a non-zero value. Either you have to ignore errors completely, or capture the output of the command and look for the "didn't contain changes" string – which is no fun at all in Bash.

I haven't been able to think of a situation where I want Cuffsert to exit with a non-zero value when there are no changes. If someone can come up with one it would probably be possible to add a flag that enabled a mode that treated no changes as an error state – but I would argue that the change I'm proposing here should be the default and that flag should be added when needed.

iconara committed:
Separate the two aspects of the does-not-update test

iconara commented on spec/cuffsert/actions_spec.rb:
```diff
@@ -292,11 +292,37 @@
+      end
+    end
+
+    it 'ends with a NoChanges action' do
```
I don't know if this test description makes sense or not.

grddev reviewed (commented):
I think the code changes make sense, but I'm not sure about the behavior change. In my manual command-line usage, no changes has mostly been because I've tried to apply an update in the wrong region or so, and then getting an error has been beneficial. Perhaps we could introduce a `--allow-empty` flag or something similar for this behavior.

iconara commented:
Ok, so you don't agree with my position that this should be the default behaviour? I can look at adding a flag, hopefully it won't be too complicated. I would prefer that the flag was something like `--expect-update` (i.e. default is that no change is business as usual) rather than `--allow-no-changes` (i.e. default is like it is now).

I'm thinking primarily of the scripting scenario, and you're talking about the interactive scenario. Perhaps it makes sense your way (i.e. `--allow-no-update`) because having to add extra flags in a scripting scenario isn't such a big deal, but requiring extra flags in interactive scenarios is annoying. The big deal for the scripting scenario is the exit code, and that shouldn't matter in an interactive scenario, right?

That being said I'm not sure that an error is what I expect in the interactive scenario either. I especially don't expect the message "Update failed: …". The change I'm proposing changes this to "No changes." (in green), which is still different than than "Done."

What about a compromise (that I implemented and reverted): zero exit code, but yellow/orange colour for the "No changes." status message?

iconara commented:
I'm trying to invite @bittrance back into this repo, but that requires him to respond to an invite. Maybe we can discuss it over lunch on Friday all of us.

grddev commented:
> That being said I'm not sure that an error is what I expect in the interactive scenario either. I especially don't expect the message "Update failed: …". The change I'm proposing changes this to "No changes." (in green), which is still different than than "Done."

Ah, sorry. I missed that the "No changes." message would actually be written instead of Done. From an interactive perspective, I would argue that the message is more important than the exit code. It is probably enough like that, but to make the difference a bit clearer, it might make sense to include the original error message (but still in green).

iconara commented:
@bittrance can you add me and @grddev as owners of the gem? `gem owner cuffsert -a iconara` & `gem owner cuffsert -a grddev`

iconara commented:
This closes https://github.com/burtcorp/cuffsert/issues/38

iconara commented:
I released this as 0.14.0, and made a fix in 0.14.1 for a thing I had missed in this PR.
